src/Components/Endpoints/src/DependencyInjection/ComponentPrerenderer.cs

```diff
@@ -2,38 +2,40 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System.Text.Encodings.Web;
-using Microsoft.AspNetCore.Components;
 using Microsoft.AspNetCore.Components.Authorization;
 using Microsoft.AspNetCore.Components.Infrastructure;
 using Microsoft.AspNetCore.Components.Routing;
 using Microsoft.AspNetCore.Components.Web;
+using Microsoft.AspNetCore.DataProtection;
 using Microsoft.AspNetCore.Html;
 using Microsoft.AspNetCore.Http;
 using Microsoft.AspNetCore.Http.Extensions;
-using Microsoft.AspNetCore.Mvc.Rendering;
 using Microsoft.Extensions.DependencyInjection;
 
-namespace Microsoft.AspNetCore.Mvc.ViewFeatures;
+namespace Microsoft.AspNetCore.Components.Endpoints;
 
 // Wraps the public HtmlRenderer APIs so that the output also gets annotated with prerendering markers.
 // This allows the prerendered content to switch later into interactive mode.
 // This class also deals with initializing the standard component DI services once per request.
-internal sealed class ComponentPrerenderer
+internal sealed class ComponentPrerenderer : IComponentPrerenderer
 {
     private static readonly object ComponentSequenceKey = new object();
     private static readonly object InvokedRenderModesKey = new object();
 
     private readonly HtmlRenderer _htmlRenderer;
-    private readonly ServerComponentSerializer _serverComponentSerializer;
+    private readonly Lazy<ServerComponentSerializer> _serverComponentSerializer;
     private readonly object _servicesInitializedLock = new();
-    private Task _servicesInitializedTask;
+    private Task? _servicesInitializedTask;
 
     public ComponentPrerenderer(
         HtmlRenderer htmlRenderer,
-        ServerComponentSerializer serverComponentSerializer)
+        IServiceProvider services)
     {
-        _serverComponentSerializer = serverComponentSerializer;
         _htmlRenderer = htmlRenderer;
+
+        // Lazy because we don't actually want to require a whole chain of services including Data Protection
+        // to be required unless you actually use Server render mode.
+        _serverComponentSerializer = new(services.GetRequiredService<ServerComponentSerializer>);
```

Comment:
FYI, I believe this introduces locking. If this is not something that needs to be accessed concurrently, there are a couple alternatives. Check the remarks [here](https://learn.microsoft.com/en-us/dotnet/api/system.lazy-1.-ctor?view=net-7.0#system-lazy-1-ctor(system-func((-0))))

If you want to remove an allocation, it's likely better that you declare an extra property and use `LazyInitializer` instead to initialize the field.

Reply:
Good point - this can be simplified and doesn't really need either `Lazy<T>` or `LazyInitializer`. Updated.